### ISSUE TITLE
Py3 Compatibility 

### ISSUE THREAD

**Issue by evanfeinberg** [evanfeinberg/conformation#4]
Significant changes are being applied to enforce Python 3 compatibility. Expect a new PR soon.

**Comment by evanfeinberg** [comment]
Notes on Python 3 compatibility: 

(1) [Easiest] print statements require parentheses. i.e., ```print "hello world"``` no longer works, requires ```print("hello world")```

(2) For comparing instances of objects, the ```__cmp__()``` function no longer suffices. Instead, you need a ```__lt__()``` method in your class. For example, in ```residue.py``` in this package:

```
  def __lt__(self, other):
    return str(self) < str(other)
```

(3) Formatting of strings has changed. If you read a file with syntax like:

```
myfile = open("my_file.txt", "rb")
for line in myfile.readlines():
  if line.split()[0] == "Hello":
    print("World")
```
This will *not* work since the string comparison will yield a Fail every time. Instead, open the file as ```open("my_file.txt", "r")```, NOT ```rb```.

(4) Python 2 pickles are NOT forward compatible with Python 3! This means if you fit models (tICA, MSMs, etc.) and saved them with the pickle/joblib protocol (e.g., with ```verbosedump```), this engenders a significant problem. A workaround that seems to suffice thus far is:

```
def compat_verboseload(filename):
  with open(filename, 'rb') as f:
    d = pickle.load(f, encoding='latin1') 
  return d
```
However, I've largely resorted to refitting all models, regardless of how cumbersome it is, to avoid having to even consider the possibility of corruption in interoperating between Py2 and Py3 pickle files.

(5) List comprehensions: In the past, I have frequently made use of nested list comprehensions. For example, using ```mdtraj``` to illustrate:

```
if "CB" in [str(a) for a in r.atoms for r in top.residues if r.index == ind_i]
```

For some reason, Python 3 really does not like this syntax. So I've replaced this line with the clumsier but equivalent:

```
          test_atoms = []
          for r in top.residues:
            if r.index != ind_j:
              continue
            test_atoms.append([str(a) for a in r.atoms])
          if "CB" in test_atoms:
            ind_j = j
```

Moral of the story: perhaps stay away from nested list comprehensions if you can in Py3. 

(6) Please add your own encounters with incompatibilities here so others will be aware for when they upgrade and/or cognizant of it so they can better future-proof their code!

**Comment by evanfeinberg** [comment]
(7) ```xrange()``` no longer works. Just use ```range()```

**Comment by rmcgibbo** [comment]
Most of the simple stuff can be fixed automatically using tools like [python-future](http://python-future.org/automatic_conversion.html)

**Comment by evanfeinberg** [comment]
Yup, ```2to3.py``` is also useful. Alas, not all stuff is simple stuff.

**Comment by mpharrigan** [comment]
(4) 
```python
pickle.dump(obj, f, protocol=2)
```

**Comment by evanfeinberg** [comment]
Is that for loading or for saving? It looks like the latter.

**Comment by mpharrigan** [comment]
It's for saving into a pickle form that can be loaded by both py2 and py3

**Comment by mpharrigan** [comment]
You can't (load in py2) a pickle (serialized in py3).
You can (load in py3) a pickle (serialized in py2)

I don't know what your code snippet is for

**Comment by evanfeinberg** [comment]
I have not found the latter to be true. Files saved either by verbosdump()
or pickle.dump() (which are essentially equivalent) in Py2 are not loadable
with verboseload() or pickle.load() in Py3, without that slight trickery.

**Comment by mpharrigan** [comment]
I just tried the following and it worked as expected

```python
Python 2.7.11 |Continuum Analytics, Inc.| (default, Dec  6 2015, 18:08:32) 
Type "copyright", "credits" or "license" for more information.

In [1]: blah = {'hi': 1, 2: [9.0, 6.0]}

In [2]: import pickle

In [3]: with open("test.pickl", 'wb') as f:
   ...:     pickle.dump(blah, f)
```

```python
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

In [1]: import pickle

In [2]: with open("test.pickl", 'rb') as f:
   ...:     blah = pickle.load(f)
   ...:     

In [3]: blah
Out[3]: {2: [9.0, 6.0], 'hi': 1}
```

**Comment by mpharrigan** [comment]
Reading about the "encoding" argument

> Optional keyword arguments are fix_imports, encoding and errors, which are used to control compatibility support for pickle stream generated by Python 2. If fix_imports is true, pickle will try to map the old Python 2 names to the new names used in Python 3. The encoding and errors tell pickle how to decode 8-bit string instances pickled by Python 2; these default to ‘ASCII’ and ‘strict’, respectively. The encoding can be ‘bytes’ to read these 8-bit string instances as bytes objects.

Maybe you have umlauts?

**Comment by evanfeinberg** [comment]
That's strange. I haven't been dabbling in Turkic languages lately, as I am
usually inclined to do. Perhaps it's something about the non-standard
objects that I'm pickling?
